### ISSUE TITLE
Broken link errors

### ISSUE THREAD

**Issue by alandefreitas** [boostorg/docca#105]
# Description

The quickbook file generated by docca has a few warnings, such as `Error: no ID for constraint linkend: "utility.ref.boost__base_from_member_lt__MemberType_&,_UniqueID__gt_.base_from_member".`.

# Preview

This [output](https://gist.github.com/alandefreitas/e438dcd487bab9365af5262f74a2fde7) generated from [these docs](https://github.com/alandefreitas/utility/tree/docca_ref).

# Replicating the issue

```bash
git clone -b docca_ref https://github.com/alandefreitas/utility.git
cd utility/doc
b2
```

I'm still coming up with a smaller example to replicate the issue.

**Comment by vinniefalco** [comment]
```
(base) alandefreitas@alandefreitass-MacBook-Pro doc % b2
Performing configuration checks

    - default address-model    : 64-bit (cached) [1]
    - default architecture     : x86 (cached) [1]
    - symlinks supported       : yes (cached)

[1] clang-12
...patience...
...found 859 targets...
...updating 6 targets...
saxonhe.saxonhe_dir ../../../bin.v2/libs/utility/doc/_reference-dir/stage1/results
saxonhe.saxonhe_dir ../../../bin.v2/libs/utility/doc/_reference-dir/stage2/results
saxonhe.saxonhe ../../../bin.v2/libs/utility/doc/reference.qbk
quickbook.quickbook-to-boostbook ../../../bin.v2/libs/utility/doc/utility_doc.xml
Generating output file: ../../../bin.v2/libs/utility/doc/utility_doc.xml
xslt-xsltproc ../../../bin.v2/libs/utility/doc/utility_doc.docbook
xslt-xsltproc-dir html/utility_HTML.manifest
Error: no ID for constraint linkend: "utility.ref.boost__base_from_member_lt__MemberType_&,_UniqueID__gt_.base_from_member".
Error: no ID for constraint linkend: "utility.ref.boost__base_from_member_lt__MemberType_&,_UniqueID__gt_.member".
Error: no ID for constraint linkend: "utility.ref.boost__basic_string_ref.operator_std__basic_string_lt__charT,_traits,_Allocator__gt_".
Error: no ID for constraint linkend: "utility.ref.boost__basic_string_view.operator_std__basic_string_lt__charT,_traits,_Allocator__gt_".
Error: no ID for constraint linkend: "utility.ref.boost__call_traits_lt__T_&__gt_.const_reference".
Error: no ID for constraint linkend: "utility.ref.boost__call_traits_lt__T_&__gt_.param_type".
Error: no ID for constraint linkend: "utility.ref.boost__call_traits_lt__T_&__gt_.reference".
Error: no ID for constraint linkend: "utility.ref.boost__call_traits_lt__T_&__gt_.value_type".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_type.overload1boost__compressed_pair.first_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_type.overload1boost__compressed_pair.first_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_type.overload1boost__compressed_pair.first_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_type.overload1boost__compressed_pair.first_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_type.overload1boost__compressed_pair.second_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_type.overload1boost__compressed_pair.second_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_type.overload1boost__compressed_pair.second_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_type.overload1boost__compressed_pair.second_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_param_type.overload1boost__compressed_pair.first_param_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_param_type.overload1boost__compressed_pair.first_param_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_param_type.overload1boost__compressed_pair.first_param_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_param_type.overload1boost__compressed_pair.first_param_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_param_type.overload1boost__compressed_pair.second_param_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_param_type.overload1boost__compressed_pair.second_param_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_param_type.overload1boost__compressed_pair.second_param_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_param_type.overload1boost__compressed_pair.second_param_type.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_reference.overload1boost__compressed_pair.first_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_reference.overload1boost__compressed_pair.first_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_reference.overload1boost__compressed_pair.first_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_reference.overload1boost__compressed_pair.first_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_reference.overload1boost__compressed_pair.second_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_reference.overload1boost__compressed_pair.second_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_reference.overload1boost__compressed_pair.second_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_reference.overload1boost__compressed_pair.second_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_const_reference.overload1boost__compressed_pair.first_const_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_const_reference.overload1boost__compressed_pair.first_const_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_const_reference.overload1boost__compressed_pair.first_const_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first_const_reference.overload1boost__compressed_pair.first_const_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_const_reference.overload1boost__compressed_pair.second_const_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_const_reference.overload1boost__compressed_pair.second_const_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_const_reference.overload1boost__compressed_pair.second_const_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second_const_reference.overload1boost__compressed_pair.second_const_reference.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.compressed_pair.overload1boost__compressed_pair.compressed_pair.overload4".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.compressed_pair.overload1boost__compressed_pair.compressed_pair.overload4".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.compressed_pair.overload2boost__compressed_pair.compressed_pair.overload5".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.compressed_pair.overload2boost__compressed_pair.compressed_pair.overload5".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.compressed_pair.overload3boost__compressed_pair.compressed_pair.overload6".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.compressed_pair.overload3boost__compressed_pair.compressed_pair.overload6".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.compressed_pair.overload1boost__compressed_pair.compressed_pair.overload4".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.compressed_pair.overload1boost__compressed_pair.compressed_pair.overload4".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.compressed_pair.overload2boost__compressed_pair.compressed_pair.overload5".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.compressed_pair.overload2boost__compressed_pair.compressed_pair.overload5".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.compressed_pair.overload3boost__compressed_pair.compressed_pair.overload6".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.compressed_pair.overload3boost__compressed_pair.compressed_pair.overload6".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first.overload1boost__compressed_pair.first.overload3".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first.overload1boost__compressed_pair.first.overload3".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first.overload2boost__compressed_pair.first.overload4".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first.overload2boost__compressed_pair.first.overload4".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first.overload1boost__compressed_pair.first.overload3".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first.overload1boost__compressed_pair.first.overload3".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first.overload2boost__compressed_pair.first.overload4".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.first.overload2boost__compressed_pair.first.overload4".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second.overload1boost__compressed_pair.second.overload3".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second.overload1boost__compressed_pair.second.overload3".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second.overload2boost__compressed_pair.second.overload4".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second.overload2boost__compressed_pair.second.overload4".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second.overload1boost__compressed_pair.second.overload3".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second.overload1boost__compressed_pair.second.overload3".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second.overload2boost__compressed_pair.second.overload4".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.second.overload2boost__compressed_pair.second.overload4".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.swap.overload1boost__compressed_pair.swap.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.swap.overload1boost__compressed_pair.swap.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.swap.overload1boost__compressed_pair.swap.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair.swap.overload1boost__compressed_pair.swap.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.first_const_reference".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.first_param_type".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.first_reference".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.first_type".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.second_const_reference".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.second_param_type".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.second_reference".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.second_type".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.compressed_pair".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.first".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.second".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.swap".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.compressed_pair.overload1".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.compressed_pair.overload1".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.compressed_pair.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.compressed_pair.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.compressed_pair.overload3".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.compressed_pair.overload3".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.first.overload1".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.first.overload1".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.first.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.first.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.second.overload1".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.second.overload1".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.second.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__compressed_pair_lt__T,_T__gt_.second.overload2".
Error: no ID for constraint linkend: "utility.ref.boost__initialized.operator_T_const_&".
Error: no ID for constraint linkend: "utility.ref.boost__initialized.operator_T&".
Error: no ID for constraint linkend: "utility.ref.boost__value_initialized.operator_T_const_&".
Error: no ID for constraint linkend: "utility.ref.boost__value_initialized.operator_T&".
Error: no ID for constraint linkend: "utility.ref.boost__basic_string_view.operator_std__basic_string_lt__charT,_traits,_Allocator__gt_".
Error: no ID for constraint linkend: "utility.ref.boost__basic_string_view.operator_std__basic_string_lt__charT,_traits,_Allocator__gt_".
Error: no ID for constraint linkend: "utility.ref.boost__basic_string_view.operator_std__basic_string_lt__charT,_traits,_Allocator__gt_".
Error: no ID for constraint linkend: "utility.ref.boost__basic_string_view.operator_std__basic_string_lt__charT,_traits,_Allocator__gt_".
Error: no ID for constraint linkend: "utility.ref.boost__basic_string_ref.operator_std__basic_string_lt__charT,_traits,_Allocator__gt_".
Error: no ID for constraint linkend: "utility.ref.boost__basic_string_ref.operator_std__basic_string_lt__charT,_traits,_Allocator__gt_".
Error: no ID for constraint linkend: "utility.ref.boost__basic_string_ref.operator_std__basic_string_lt__charT,_traits,_Allocator__gt_".
Error: no ID for constraint linkend: "utility.ref.boost__basic_string_ref.operator_std__basic_string_lt__charT,_traits,_Allocator__gt_".
...updated 6 targets...
(base) alandefreitas@alandefreitass-MacBook-Pro doc % 
```

**Comment by evanlenz** [comment]
Looking more closely at this, these are actually not Quickbook warnings. (We do have issue #76 regarding some Quickbook warnings.)

The above errors are emitted by the DocBook stylesheets signifying broken links. I believe we're generating some invalid links like this:

![image](https://user-images.githubusercontent.com/604974/138372382-98c4ca72-8a37-4ec1-af08-a32ea59eb8a9.png)

I believe the comma in the ID is causing the problem. (The DocBook stylesheets apparently don't like this.)

Adding this to common.xsl fixed the few examples I reproduced. I'll have to check to see if there are other characters causing the problem (e.g. ampersands):

```xml
     <replace pattern="," with="_comma_"/>
```

**Comment by alandefreitas** [comment]
Some good news! 🎉

I've managed to put the [reference in a child section](https://alandefreitas.github.io/boost-utility-docs/index.html) so that the TOC looks cleaner while fixing [the links in the subsection](https://alandefreitas.github.io/boost-utility-docs/utility/index_and_ref/ref.html) and getting the up arrows to work. The up arrows are not going to any blank pages anymore. For instance, the up arrow in [this page](https://alandefreitas.github.io/boost-utility-docs/utility/index_and_ref/ref/boost__result_of.html) goes [back to the reference](https://alandefreitas.github.io/boost-utility-docs/utility/index_and_ref/ref.html). 

It's looking quite good. We can even reuse this pattern in Boost.Json to remove that blank page from there. 

We only have [a few broken links](https://gist.github.com/alandefreitas/35fa2fa10f158e1bf9725b2112f3fdab) now, but they are in two groups and we know what they are:

1. The links that contain an ampersand `&` in them, such as `utility.index_and_ref.ref.boost__call_traits_lt__T_&__gt_.const_reference`
2. The links for the compressed_pair "false overloads", such as `utility.index_and_ref.ref.boost__compressed_pair.second.overload1boost__compressed_pair.second.overload3`

I think @evanlenz already identified a fix for both of them. And I'll also add some macros to remove whatever is in `ob_compressed_pair.hpp` for redundancy.

**Comment by alandefreitas** [comment]
More good news. No warnings left. [The result](https://alandefreitas.github.io/boost-utility-docs/utility/index_and_ref/ref.html) looks good. All links are OK.

```
Performing configuration checks

    - default address-model    : 64-bit (cached) [1]
    - default architecture     : x86 (cached) [1]
    - symlinks supported       : yes (cached)

[1] clang-12
...patience...
...found 859 targets...
...updating 4 targets...
saxonhe.saxonhe ../../../bin.v2/libs/utility/doc/reference.qbk
quickbook.quickbook-to-boostbook ../../../bin.v2/libs/utility/doc/utility_doc.xml
Generating output file: ../../../bin.v2/libs/utility/doc/utility_doc.xml
xslt-xsltproc ../../../bin.v2/libs/utility/doc/utility_doc.docbook
xslt-xsltproc-dir html/utility_HTML.manifest
...updated 4 targets...
```

The version without the warnings will only be correct in the preview because it required some local changes. We need some PRs to fix it permanently.

1.

That required

```xml
<replace pattern="," with="_comma_"/>
<replace pattern="&amp;" with="_ampersand_"/>
```

in docca, so we need a PR for that. It's probably worth going through all HTML escaped chars that might cause errors and putting them there already.

2.

The compressed_pair duplicates were manually removed with macros for now. So we need one more PR for that.

**Comment by evanlenz** [comment]
Excellent work! I am still working on the PR for all the broken-link issues. That includes the change you listed above, the treatment of friend sections as always public, and the duplicates removal. All are implemented and now I'm just verifying the changes across all the libraries to guard against regressions. (I suspect the bad-filename issue in the utility project will persist for me, but that sounds unique to my environment and I will have to address that separately.) Stay tuned...

**Comment by evanlenz** [comment]
In order to ensure the bulk of this work isn't lost, I'm going to revisit, verify, and commit the relevant changes thus far. We can open a new issue to address any remaining broken links we might come across.

**Comment by sdkrystian** [comment]
Fixed by 8ba36c0c1e1adb32dbf4cd9c7f3cb36f339b8b8d
